### ISSUE TITLE
Inconsistent partition IDs for different dataset splits using NaturalIdPartitioner

### ISSUE THREAD

**Issue by adamtupper** [adap/flower#5243]
### Describe the bug

The `NaturalIdPartitioner` creates natural IDs using `Dataset.unique()`, where the ordering of the returned IDs depends on the order of examples in the dataset. Then, the partition IDs are assigned based on the of the natural IDs:

```python
def _create_int_partition_id_to_natural_id(self) -> None:
    """Create a mapping from int indices to unique client ids from dataset.

    Natural ids come from the column specified in `partition_by`.
    """
    unique_natural_ids = self.dataset.unique(self._partition_by)
    self._partition_id_to_natural_id = dict(
        zip(range(len(unique_natural_ids)), unique_natural_ids)
    )
```

This is problematic when creating multiple partitioners (e.g., for training and test splits) as the mappings from partition ID to natural ID can be different for each. In this scenario, if the mappings are not the same the clients will be trained and tested on on different _natural_ partitions (even though the partition IDs are the same).

This could be fixed by sorting the unique natural IDs before assigning the partition IDs:

```python
def _create_int_partition_id_to_natural_id(self) -> None:
    """Create a mapping from int indices to unique client ids from dataset.

    Natural ids come from the column specified in `partition_by`.
    """
    unique_natural_ids = sorted(self.dataset.unique(self._partition_by))
    self._partition_id_to_natural_id = dict(
        zip(range(len(unique_natural_ids)), unique_natural_ids)
    )
```

### Steps/Code to Reproduce

```python
from datasets import Dataset
from flwr_datasets.partitioner import NaturalIdPartitioner

# Create datasets
train_data = {
    "features": [1, 2, 3],
    "labels": [0, 0, 1],
    "clients": ["a", "b", "a"],
}
test_data = {
    "features": [4, 5, 6],
    "labels": [1, 0, 0],
    "clients": ["b", "a", "a"],
}
train_dataset = Dataset.from_dict(train_data)
test_dataset = Dataset.from_dict(test_data)

# Create partitioners
train_partitioner = NaturalIdPartitioner(partition_by="clients")
test_partitioner = NaturalIdPartitioner(partition_by="clients")
train_partitioner.dataset = train_dataset
test_partitioner.dataset = test_dataset

# Check partitioner mappings
train_partitioner.load_partition(partition_id=0)
print("Train partition mapping", train_partitioner.partition_id_to_natural_id)

test_partitioner.load_partition(partition_id=0)
print("Test partition mapping", test_partitioner.partition_id_to_natural_id)

```

### Expected Results

```shell
Generating partition_id_to_indices: 3it [00:00, 91180.52it/s]
Train partition mapping {0: 'a', 1: 'b'}
Generating partition_id_to_indices: 3it [00:00, 96791.63it/s]
Test partition mapping {0: 'a', 1: 'b'}
```

### Actual Results

```shell
Generating partition_id_to_indices: 3it [00:00, 91180.52it/s]
Train partition mapping {0: 'a', 1: 'b'}
Generating partition_id_to_indices: 3it [00:00, 96791.63it/s]
Test partition mapping {0: 'b', 1: 'a'}
```

**Comment by jafermarq** [comment]
@adamtupper many thanks for creating this issue and providing all this context and info. We'll look into it soon

**Comment by adamtupper** [comment]
You're welcome! ☺️ Let me know if you'd like me to open a pull request to fix it, otherwise I'll leave it to the team to decide the best way forward.

**Comment by WilliamLindskog** [comment]
Hi @adamtupper, 

Thank you for your explanation. It really helps understanding it. Would you be interested in opening up an PR for this? 

Best regards
William

**Comment by WilliamLindskog** [comment]
Awesome! Let me know if there's anything I can help with 👍🏽

**Comment by adamtupper** [comment]
Thanks for your patience!

I've discovered a similar issue with the `DirichletPartitioner` that results in train/test splits having different label distributions for the same client due to the same ordering issue.

I'll submit a pull request to fix both later today. 😄

**Comment by adamtupper** [comment]
Hi @WilliamLindskog,

I've submitted a PR to address this issue for both the `NaturalIdParitioner` and `DirichletPartitioner`. Let me know if there's anything you or the team would like to discuss or change!

Cheers,

Adam

**Comment by WilliamLindskog** [comment]
Hi @adamtupper, 

Thanks for notifying us. I think you and @jafermarq are already discussing steps forward in PR so I'll leave it to you to sort out. We can close this issue after this is merged. 

Best regards
William
